### PR TITLE
internal/docstore: preliminary dynamo queries

### PR DESCRIPTION
**Pull request by shantuo**
This is still WIP:

- Partition key is hard-coded when doing query since we can only have one `Where` call.
- The table schema is changed to include partition and sort key.
- When sort key is defined, it cannot be blank, this fails the unnamed put tests, commented out for now.

### CODE REVIEW

**Comment by jba** [comment]
I'm confused about why we have to change the conformance tests to have a Kind field. What if instead the dynamo driver automatically inserted some fixed value for the partition key for every action? That way the other drivers and the conformance tests wouldn't need to know about it.

**Comment by shantuo** [comment]
I made the choice of changing the tests since I thought those are easier to change without breaking what we had. I see the fs client is going to support multiple filters, maybe we can wait until your PR lands so that I can continue on this one? Is there anything else blocking?

File: internal/docstore/dynamodocstore/query.go [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+}
+
+func (it *documentIterator) Next(ctx context.Context, doc driver.Document) error {
+	if it.limit > 0 && it.curr >= it.limit ||
+		it.curr >= len(it.items) && it.last == nil {
+		return io.EOF
+	}
+	if it.curr >= len(it.items) && it.last != nil {
+		// Make a new query request at the end of this page.
+		i, err := it.c.runGetQuery(ctx, it.q, it.last)
```
Comment (jba) [inline internal/docstore/dynamodocstore/query.go]: It seems odd to me to create a new iterator and then copy from it. How about if runGetQuery returns the query output struct (what you call `out` above), and you use that to populate `it` here?

**Comment by jba** [comment]
> maybe we can wait until your PR lands so that I can continue on this one

No, what you have is fine. Let's continue without waiting.

File: internal/docstore/dynamodocstore/query.go [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+		c:     c,
+		q:     q,
+		items: out.Items,
+		count: *out.Count,
+		limit: q.Limit, // manually count limit since dynamodb uses "limit" as scan limit before filtering
+		last:  out.LastEvaluatedKey,
+	}, nil
+}
+
+func processFilter(cb expression.Builder, fs []driver.Filter, pkey, skey string) expression.Builder {
```
[the remark this answers is not in the document]
Comment (shantuo) [reply internal/docstore/dynamodocstore/query.go]: Done.

File: internal/docstore/dynamodocstore/query.go [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+			return expression.KeyLessThan(key, val), true
+		case "<=":
+			return expression.KeyLessThanEqual(key, val), true
+		case "=":
+			return expression.KeyEqual(key, val), true
+		case ">=":
+			return expression.KeyGreaterThanEqual(key, val), true
+		case ">":
+			return expression.KeyGreaterThan(key, val), true
+		}
```
[the remark this answers is not in the document]
Comment (shantuo) [reply internal/docstore/dynamodocstore/query.go]: Done.

File: internal/docstore/dynamodocstore/query.go [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+		return expression.LessThan(name, val)
+	case "<=":
+		return expression.LessThanEqual(name, val)
+	case "=":
+		return expression.Equal(name, val)
+	case ">=":
+		return expression.GreaterThanEqual(name, val)
+	case ">":
+		return expression.GreaterThan(name, val)
+	}
```
[the remark this answers is not in the document]
Comment (shantuo) [reply internal/docstore/dynamodocstore/query.go]: Done.

File: internal/docstore/dynamodocstore/query.go (hunk 1 above)
Comment (shantuo) [reply internal/docstore/dynamodocstore/query.go]: Done.

**Comment by jba** [comment]
This test is failing:
```
--- FAIL: TestConformance (0.16s)
    --- FAIL: TestConformance/Query (0.03s)
        replay.go:71: Replaying from golden file testdata/TestConformance/Query
        replay.go:119: matched request #0 (POST https://dynamodb.us-east-2.amazonaws.com/)
        replay.go:119: matched request #1 (POST https://dynamodb.us-east-2.amazonaws.com/)
        replay.go:115: mismatched HTTP body at request #2
FAIL
coverage: 10.5% of statements in ./...
FAIL	gocloud.dev/internal/docstore/dynamodocstore	0.260s
```

**Comment by shantuo** [comment]
Oh, they fail because I only re-recorded the query one.

**Pull request merged**